### PR TITLE
Production scripts for flattening and concatenating CAFs

### PR DESCRIPTION
**Pull request by etyley**

### CODE REVIEW

File: scripts/poms/sbndpoms_caf_concatenation.sh [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+  case $1 in
+    -h|-\?|--help)
+      show_help    # Display a usage synopsis.
+      exit
+      ;;
+    --defname)       # Takes an option argument; ensure it has been specified.
+      if [ "$2" ]
+      then
+        DEFNAME="$2"
+        shift
```
Comment (cjbacchus) [inline scripts/poms/sbndpoms_caf_concatenation.sh]: shouldn't this be `shift 2`? And ditto for workdir, outdir, maxsize below.

File: scripts/poms/sbndpoms_caf_concatenation.sh [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+echo "  Max Size: $(numfmt --to=si $MAXSIZE)"
+if [ ! -z "$FLATTEN" ]
+then
+  echo "  Making Flat CAFS"
+fi
+
+prepare()
+{
+  # Assume that the metadata is roughly the same in all files, so take the first
+  FILENAME=$(samweb -e sbnd list-definition-files $DEFNAME | head -n 1)
```
Comment (cjbacchus) [inline scripts/poms/sbndpoms_caf_concatenation.sh]: is `samweb list-files defname: $DEFNAME --limit 1` faster? samweb might seem fast now, but we have some terrible performance on NOvA.

File: scripts/poms/sbndpoms_caf_concatenation.sh [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+  samweb -e sbnd get-metadata $FILENAME --json > base.json
+
+  # Lets change "caf" to "concat_caf" for the stage and defname
+  MDPRODUCTIONDEFNAME=$(echo "$DEFNAME" | sed "s/caf/concat_caf/g")
+
+  # Now we want to set some global vars to define the ouput path
+  MDFILETYPE=$(cat base.json | grep "file_type" | cut -d\" -f4)
```
Comment (cjbacchus) [inline scripts/poms/sbndpoms_caf_concatenation.sh]: Nitpick: these can all be `grep base.json "something"`
Potentially `jq` is a better way to deal with json on the command line.
Or take a look at `/cvmfs/nova.opensciencegrid.org/externals/novaproduction/v05.22/NULL/bin/get_metadata_field` (it's a python script). Maybe we want something like that?

File: scripts/poms/sbndpoms_caf_concatenation.sh [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+
+  if [ -z "$(ls -A $OUTDIR)" ]
+  then
+    echo "OUTDIR: $OUTDIR"
+  else
+    echo "Not Empty: $OUTDIR"
+    exit 3
+  fi
+
+  if [ 0 -ne $(samweb -e sbnd list-definitions | grep "$MDPRODUCTIONDEFNAME" | wc -l) ]
```
Comment (cjbacchus) [inline scripts/poms/sbndpoms_caf_concatenation.sh]: `grep` already returns a success/failure code. I think you can do `if samweb -e sbnd list-definitions | grep -q $MDPRODUCTIONDEFNAME`

File: scripts/poms/sbndpoms_caf_concatenation.sh [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+  then
+    echo "SAM Definition $MDPRODUCTIONDEFNAME already present"
+    exit 3
+  else
+    echo "Creating SAM Defintion $MDPRODUCTIONDEFNAME"
+  fi
+
+  if [ ! -z "$FLATTEN" ]
+  then
+    FLATMDPRODUCTIONDEFNAME=$(echo "$MDPRODUCTIONDEFNAME" | sed "s/concat_caf/flat_caf/g")
```
Comment (cjbacchus) [inline scripts/poms/sbndpoms_caf_concatenation.sh]: More bash nitpicking. You can write `FLATMDPRODUCTIONDEFNAME=${MDPRODUCTIONDEFNAME/concat_caf/flat_caf}`. I admit I'm not sure whether this does/doesn't imply `/g`

File: scripts/poms/sbndpoms_caf_concatenation.sh [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+    samweb -e sbnd add-file-location ${UNIQUEOUTFLATNAME} $OUTDIR
+    echo "$UNIQUEOUTFLATNAME declared and located by SAM"
+  fi
+}
+
+prepare
+
+# Loop over all of the files in the dataset
+DEFSIZE=$(samweb -e sbnd list-definition-files --summary $DEFNAME | grep "Total size" | tr -dc '0-9')
+FILECOUNT=$(samweb -e sbnd list-definition-files --summary $DEFNAME | grep "File count" | tr -dc '0-9')
```
Comment (cjbacchus) [inline scripts/poms/sbndpoms_caf_concatenation.sh]: `samweb count-definition-files`. Not sure I have an alternative for the size though.

File: scripts/poms/sbndpoms_caf_concatenation.sh [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+  SLICEDEFNAME=${DEFNAME}_Slice${SLICENUM}
+  samweb -e sbnd create-definition "$SLICEDEFNAME" "defname: $DEFNAME with limit $FILELIMIT with offset $FILECOUNTER"
+
+  doConcat
+
+  SLICENUM+=1
+done
+
+# Create a definition with the output files
+samweb -e sbnd create-definition $MDPRODUCTIONDEFNAME "file_name like concat_caf_%.root and file_type $MDFILETYPE and production.type $MDPRODUCTIONTYPE and production.name $MDPRODUCTIONNAME and sbnd_project.name $MDSBNDPROJECTNAME and sbnd_project.version $MDSBNDPROJECTVERSION and sbnd_project.stage $MDSBNDPROJECTSTAGE"
```
Comment (cjbacchus) [inline scripts/poms/sbndpoms_caf_concatenation.sh]: You could throw in `is_child_of: ( defname: $DEFNAME )` (not 100% sure of the syntax). I'm kind of paranoid about this sort of definition going wrong.

**Review by cjbacchus — COMMENTED** [review]
Lots of nitpicking about samweb/bash syntax, a few concerns about robustness. None of this is critical. I also made a few comments that aren't included in this review, because I can't use the github interface properly.

**Comment by etyley** [comment]
Thanks for your input on this. I have implemented all of your suggestions and checked that everything still works as expected

**Comment by cjbacchus** [comment]
Thanks. How about the very first comment with `shift`? I think you are leaving the $2 argument in the command-line to confuse future passes through the loop. Or have I misunderstood something?

**Comment by etyley** [comment]
Changing the `shift` to a `shift 2` stopped it parsing the `--flatten` argument properly. I was going to go back to it once I implemented the other changes but forgot. I will look into it now

**Comment by etyley** [comment]
There is a second shift on line 62 which was causing it to skip the `--flatten` when making your suggested change. I have removed that and it now works

**Comment by cjbacchus** [comment]
Makes sense. I missed that other `shift`, but I think that meant that `--flatten` was actually removing two arguments previously (one itself and one at the end of the loop). In any case, it all looks good now. Thanks!

**Review by cjbacchus — APPROVED** [review]

**Pull request merged**